server/db/models/campaigns.js

```diff
@@ -10,5 +11,76 @@ export default db.Model.extend({
   },
   calls() {
     return this.hasMany(Call);
+  },
+  responses() {
+    return this.hasMany(Response).through(Call);
+  },
+  getExport() {
+    // getExport should only be called on campaigns that were loaded with related calls.responses
+    const calls = this.relations.calls;
+
+    // some records have duplicate props, this renames the props and returns
+    // a version of the object that no longer has the duplicate prop name.
+    function samePropProcessing(object, objectType, propName, index) {
+      // create a copy because linter likes functional
+      const cloneObject = object;
+      if (index) {
+        // if we're given an index, it's for naming purposes, use it
+        cloneObject[`${objectType}_${index}_${propName}`] = cloneObject[propName];
+      } else {
+        cloneObject[`${objectType}_${propName}`] = cloneObject[propName];
+      }
+      // delete the potential duplicate prop name
+      delete cloneObject[propName];
+      return cloneObject;
+    }
+    // this will be an accumulator that will collect csv headers from record keys
+    let headers = [];
+    // make an array of strings that are lines in the csv.
+    const joined = calls.map((call) => {
+      const responses = call.relations.responses;
+      // this variable will hold the complete record for one line in the csv
+      // but we want to set it conditionally
+      let record;
+      // reassign the cloned call with each processing to keep it clean as we go.
+      let callClone = samePropProcessing(call.attributes, 'call', 'id');
+      callClone = samePropProcessing(callClone, 'call', 'updated_at');
+      callClone = samePropProcessing(callClone, 'call', 'created_at');
+
+      if (responses.length > 0) {
+        // make an array of processed response objects from our array of response models.
+        const responsesClone = responses.models.map((response, index) => {
+          let responseClone = samePropProcessing(response.attributes, 'response', 'id', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'updated_at', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'created_at', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'response', index + 1);
+          responseClone = samePropProcessing(responseClone, 'response', 'question_id', index + 1);
+          // since these will be merged with a call object, we'll already have call_id
+          delete responseClone.call_id;
+          return responseClone;
+        });
+        // add the cloned call to the front of the array before we pass the
+        // array as a set of arguments
+        responsesClone.unshift(callClone);
+        // set our record to a merged version of the processed call and its associated responses.
+        record = Object.assign.apply(null, responsesClone);
```

Comment:
I might be missing some of the underlying complexity here, but here and on like 73 did you consider using a native array method?
This method copies all enumerable properties of an object, which for arrays can have unexpected results. (same reason why with arrays it's generally advised not to use for...in loops).

Reply:
Also, I think record can be assigned here, I'm not sure Object.assign is necessary. But again I could be missing some underlying complexity.

Reply:
Object.assign is a good ES6 way to merge objects and `.apply` here calls Object.assign and there are two arguments. First defines what `this` within the method `.apply` is being called on would bind to. `null` in this case because we don't care. The second argument takes an array and passes that array's elements as arguments to the method `.apply` is called on. So here we're not operating on the array itself but the (non-array) objects within this array.

Reply:
gotcha